### PR TITLE
MGDCTRS-2024: Support Smart Event Processors in the Control Plane

### PR DESCRIPTION
**Pull request by manstis**
See [https://issues.redhat.com/browse/MGDCTRS-2024](https://issues.redhat.com/browse/MGDCTRS-2024)

## Description
Add support for Smart Event `Processors` to `cos-fleet-manager`

## Verification Steps

1. Fetch this PR
2. Fetch `cos-fleet-manager`
3. Patch `cos-fleet-manager` to use this version of `kas-fleet-manager` by adding this to its `go.mod` file:
`replace github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager v0.0.0-20230120171214-d0c418df2eee => ../kas-fleet-manager`
4. Run `cos-fleet-manager`
5. Make a Cluster
6. Make a Namespace
7. Run `make run/docs` in `kas-fleet-manager`
8. See the new `Processor` endpoints. `Get`, `List`, `Create` and `Delete` are supported up to and including the database.

## Checklist (Definition of Done)
- [ ] All acceptance criteria specified in JIRA have been completed
- [ ] Unit and integration tests added that prove the fix is effective or the feature works (tested against emulated and non-emulated OCM environment)
- [ ] ~Documentation added for the feature~
- [ ] CI and all relevant tests are passing
- [ ] Code Review completed
- [ ] Verified independently by reviewer
- [ ] All PR comments are resolved either by addressing them or creating follow up tasks
- [ ] ~Required metrics/dashboards/alerts have been added (or PR created)~
- [ ] ~Required Standard Operating Procedure (SOP) is added~
- [ ] JIRA has been created for changes required on the client side

### CODE REVIEW

**Comment by manstis** [comment]
@robpblake @danielezonca @lucamolteni @vaibhavjainwiz This is pretty ready for review now.

@lburgazzoli @valdar IDK whether somebody familiar with RHOC CP should review?

**Comment by vaibhavjainwiz** [comment]
AFAIK [kas-fleet-manager](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager) is shared codebase between RHOSAK and RHOC. The processor functionality is required just for `RHOC`. I think Processor related code should be in [cos-fleet-manager](https://github.com/bf2fc6cc711aee1a0c2a/cos-fleet-manager) repo

**Comment by lucamolteni** [comment]
> AFAIK [kas-fleet-manager](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager) is shared codebase between RHOSAK and RHOC. The processor functionality is required just for `RHOC`. I think Processor related code should be in [cos-fleet-manager](https://github.com/bf2fc6cc711aee1a0c2a/cos-fleet-manager) repo

A proper separation between the two would be nice to have but currently we don't have it, all RHOC code is still in `kas-fleet-manager`. I wouldn't start from this to separate the twos 

@lburgazzoli  see ^ as well

File: internal/connector/internal/api/dbapi/processor.go
Comment (robpblake) [inline internal/connector/internal/api/dbapi/processor.go:48]: @manstis Apologies if we have already discussed this, but do you know what is the status of support for `channel` on Connector and if it's something that is being actively used?

File: internal/connector/internal/api/dbapi/processor.go
Comment (robpblake) [inline internal/connector/internal/api/dbapi/processor.go:51]: I'm not sure I understand how `CloudProvider`, `Region` and `MultiAZ` work for Processors 🤔  The `namespaceId` will reference a Connector Cluster that is pre-deployed by the user, so at least the `CloudProvider` and `Region` are already set and essentially inherited from the underlying OSD on which the namespace is "hosted".

`MultiAZ` will again depend on the capabilities of the underlying OSD hosting the namespace and FWICS, we don't actually have that information about the Connector Cluster (I could be completely missing something as I'm just getting used to Go and this codebase).

File: internal/connector/internal/api/private/api_connector_clusters_agent.go
Comment (robpblake) [inline internal/connector/internal/api/private/api_connector_clusters_agent.go:511]: @manstis What's the use-case for retrieving a specific `ProcessorDeployment` by `connectorClusterId` and `processorDeploymentId`?

File: internal/connector/internal/handlers/processor_secrets.go
Comment (robpblake) [inline internal/connector/internal/handlers/processor_secrets.go:10]: @manstis Should this be `v2Alpha1` to align with the path under which we are exposing the Processor API?

File: internal/connector/internal/api/dbapi/processor.go
Comment (manstis) [reply internal/connector/internal/api/dbapi/processor.go:48]: Support for different Connector `channel`'s was added recently [here](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/commit/6e3c3d817dc4b4dab4a95937b877722bc2c637ae). I don't know how actively used it is. We could choose to remove this field for `Processor`'s (for now) but I suspect it may follow.

File: internal/connector/internal/api/dbapi/processor.go
Comment (manstis) [reply internal/connector/internal/api/dbapi/processor.go:51]: I can remove them. I will wait for all feedback to be assembled before making changes.

File: internal/connector/internal/handlers/processors.go
Comment (robpblake) [inline internal/connector/internal/handlers/processors.go:265]: @manstis Out of curiosity, how much work would it be to expose this under the admin API? That would give us the ability to stop/delete Processors that are behaving badly.

File: internal/connector/internal/api/private/api_connector_clusters_agent.go
Comment (manstis) [reply internal/connector/internal/api/private/api_connector_clusters_agent.go:511]: `Processors` are deployed to the same `namespace`/`cluster` as the `Connector`'s to which they can integrate.

There is an API to retrieve `Connector`'s by `clusterId` so I thought it *might* be useful for `Processor`'s too.

I can remove every such method if the generated (app-services) API is out of scope (leaving only the methods needed for the Data Plane/Control Plane sync). Happy to do so just let me know your thoughts.

File: internal/connector/internal/handlers/processor_secrets.go
Comment (manstis) [reply internal/connector/internal/handlers/processor_secrets.go:10]: Yes it should. Good spot! I will change.

File: internal/connector/internal/api/private/api_connector_clusters_agent.go
Comment (robpblake) [reply internal/connector/internal/api/private/api_connector_clusters_agent.go:511]:
```
Processors are deployed to the same namespace/cluster as the Connector's to which they can integrate.
```
Do you mean "can be" deployed to the same namespace/cluster? Kafka is the intermediary, so there is nothing from stopping a user from creating a `namespace`  on a completely different Connector Cluster and deploying their `Processors` there. As long as the Processor produces/consumes to the correct Kafka topics, Connectors running in completely different Connector clusters can also consume/produce to the same topics and things should "just work"™. 

```
Happy to do so just let me know your thoughts.
```

I'm sure there must be a use-case for the corresponding Connector method, but I'm not sure what it is yet. Perhaps lets check the synchronizer code to see what it's used for.

File: internal/connector/internal/config/processors.go
Comment (robpblake) [inline internal/connector/internal/config/processors.go:25]: @manstis We'll need some form of configuration for Processors in terms of revision, container image and perhaps channels. We've spoken briefly in the chat about this, but I think exact structure will depend on whether we see value in supporting `channel` for Processors from the start. Happy to sync offline about the structure so that we can make a final decision 👍

File: internal/connector/internal/api/private/api_connector_clusters_agent.go
Comment (manstis) [reply internal/connector/internal/api/private/api_connector_clusters_agent.go:511]: > I'm sure there must be a use-case for the corresponding Connector method, but I'm not sure what it is yet. 

The generated Connectors method is not used in the RHOC codebase.

IDK if there's another repository with a CLI client?

Comment (manstis) [reply internal/connector/internal/api/private/api_connector_clusters_agent.go:511]: It's not used by {{cos-tools}} (the repository for `rhoc cli`) or `cos-fleetshard`... 

I think it's safe to say it might be redundant.

File: internal/connector/internal/api/dbapi/processor.go
Comment (danielezonca) [reply internal/connector/internal/api/dbapi/processor.go:51]: I agree that `CloudProvier`, `Region` and `MultiAZ` are tied to the namespace/OSD so we can remove them from here

File: internal/connector/internal/handlers/processors.go
Comment (manstis) [reply internal/connector/internal/handlers/processors.go:265]: I'd be tempted to avoid scope creep to be honest.

These PRs are large enough already. Create a JIRA to consider a _consolidated_ Admin API might be a better option.

File: internal/connector/internal/api/dbapi/processor.go
Comment (manstis) [reply internal/connector/internal/api/dbapi/processor.go]: Removed.

File: internal/connector/internal/api/private/api_connector_clusters_agent.go
Comment (lburgazzoli) [reply internal/connector/internal/api/private/api_connector_clusters_agent.go:511]: I'm nut 100% sure as naming is not the same but for connectors, there is a rhoc subcommand to get a deployment by id: https://github.com/bf2fc6cc711aee1a0c2a/cos-tools/blob/main/rhoc/pkg/cmd/deployments/get/get.go

File: internal/connector/internal/config/processors.go
Comment (lburgazzoli) [reply internal/connector/internal/config/processors.go:25]: channels, images and co will come from a catalog at some point right ?

File: internal/connector/internal/api/dbapi/processor.go
Comment (lburgazzoli) [reply internal/connector/internal/api/dbapi/processor.go:48]: yes they are used now, primary as a gate for filtering out connectors depending on the maturity level so i.e. come connectors won't be available in prod but they will be in dev/staging. Same thinking can be applied to preview conenctors at some point

File: internal/connector/internal/config/processors.go
Comment (robpblake) [reply internal/connector/internal/config/processors.go:25]: @lburgazzoli Yep, that's the plan 👍 We're trying to be as closely aligned to the Connector catalog concept as possible, even though we only have a single Processor type right now.

File: internal/connector/internal/handlers/processors.go
Comment (robpblake) [reply internal/connector/internal/handlers/processors.go:265]: Sure thing 👍

File: internal/connector/internal/handlers/processor_secrets.go
Comment (ziccardi) [inline internal/connector/internal/handlers/processor_secrets.go:36]: This function never returns errors

File: internal/connector/internal/services/processor_deployment.go
Comment (ziccardi) [inline internal/connector/internal/services/processor_deployment.go:153]:
```suggestion
	if !shared.StringEmpty(listArgs.Search) {
```

File: internal/connector/internal/api/private/api_connector_clusters_agent.go
Comment (manstis) [reply internal/connector/internal/api/private/api_connector_clusters_agent.go:511]: See https://issues.redhat.com/browse/MGDCTRS-2269

File: internal/connector/internal/handlers/annotations.go
Comment (ziccardi) [inline internal/connector/internal/handlers/annotations.go:37]:
```suggestion
		if !arrays.IsEmpty(errs) {
```

File: internal/connector/internal/handlers/annotations.go
Comment (ziccardi) [inline internal/connector/internal/handlers/annotations.go:41]:
```suggestion
		if !arrays.IsEmpty(errs) {
```

File: internal/connector/internal/handlers/annotations.go
Comment (ziccardi) [inline internal/connector/internal/handlers/annotations.go:48]:
```suggestion
		if arrays.AnyMatch(reservedDomains, func(d string) bool { return strings.Contains(k, d)}) {
			return errors.BadRequest("cannot use reserved annotation %s from domain %s", k, d)
		}
```

File: internal/connector/internal/handlers/processor_secrets.go
Comment (manstis) [reply internal/connector/internal/handlers/processor_secrets.go:36]: Yes, this is true (and also true of the method [`stripProcessorSecretReferences(..)`](https://github.com/bf2fc6cc711aee1a0c2a/kas-fleet-manager/pull/1686/files#diff-b835efb128f9125006ba8a2c048a61d629283d3f0727169a30574dbc0fdb3bd5R12)).

At this point in time the definition of a `Processor` does not contain secret information (or better said we don't know if it will need to contain secret information) and therefore the two methods here are very basic. However if we need to make the secret handling more complicated it is probable these methods could need to return errors in the future.

I am therefore "future proofing" the code. Less so in this file itself; but in the places where these methods are invoked and we need to handle errors if/should they arise (in the future).

File: internal/connector/internal/handlers/processor_secrets.go
Comment (ziccardi) [inline internal/connector/internal/handlers/processor_secrets.go:21]:
```suggestion
	if !shared.StringEmpty(resource.ServiceAccount.ClientSecret) {
```

File: internal/connector/internal/handlers/processor_secrets_test.go
Comment (ziccardi) [inline internal/connector/internal/handlers/processor_secrets_test.go:24]:
```suggestion
	g.Expect(resource.ServiceAccount.ClientSecret).To(gomega.BeEmpty())
	g.Expect(resource.ServiceAccount.ClientSecretRef).To(gomega.BeEmpty())
```

File: internal/connector/internal/handlers/processor_secrets.go
Comment (ziccardi) [reply internal/connector/internal/handlers/processor_secrets.go:36]: In this case I would avoid a `naked return` and would use an explicit one to make it clear that we are always returning `nil`:
```go
return result, nil
```

File: internal/connector/internal/handlers/annotations.go
Comment (manstis) [reply internal/connector/internal/handlers/annotations.go:48]: This does not work as we loose visibility of the `d` parameter for the error message.

In your proposal `d` is scoped to the anonymous function.

Comment (ziccardi) [reply internal/connector/internal/handlers/annotations.go:49]: We could use this in this case:
```go
		if idx, d := arrays.FindFirst(reservedDomains, func(d string) bool { return strings.Contains(k, d) }); idx != -1 {
			return errors.BadRequest("cannot use reserved annotation %s from domain %s", k, d)
		}
```

Comment (manstis) [reply internal/connector/internal/handlers/annotations.go:49]: We could do; however I'm reluctant to change the way we iterate a collection when there appears to be no real value added? Both ultimately use `strings.Contains(k, d)` and the implementation of `arrays.FindFirst(..)` uses `range` anyway too. Is there a compelling reason to make this change?

Comment (ziccardi) [reply internal/connector/internal/handlers/annotations.go:49]: The only reason I see is that `FindFirst` makes the loop's intent clear at first sight.
But I'm happy to stick with the range loop.

**Comment by manstis** [comment]
@lburgazzoli Another `token` related issue:
```
[2023-04-13T15:24:49.104Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
[2023-04-13T15:24:49.105Z] ['verbose'] The error stack is: Error: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 404 - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
    at main (/snapshot/repo/dist/src/index.js)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
[2023-04-13T15:24:49.105Z] ['verbose'] End of uploader: 750 milliseconds
Error: Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255
```

**Review by sutaakar — APPROVED** [review]
Looks good to me (was checking mostly tests)

File: internal/connector/internal/config/processors.go
Comment (manstis) [reply internal/connector/internal/config/processors.go]: Changes made to import a "Processor Catalog" using the same approach as taken by Connectors.

**Comment by manstis** [comment]
@danielezonca @ziccardi @robpblake @lburgazzoli 

All changes emanating from your first reviews are complete.

The PR has as good as (if not better) test coverage than what exists for Connectors.

Could you please take another look so we can consider this for merging soon?

File: test/cucumber/http_request.go
Comment (manstis) [inline test/cucumber/http_request.go:255]: If an assertion timed out it used to pass.. it will now fail as timed out.

File: internal/connector/internal/handlers/processors.go
Comment (manstis) [reply internal/connector/internal/handlers/processors.go:268]: See https://issues.redhat.com/browse/MGDCTRS-2291

**Comment by manstis** [comment]
RHOC has been cancelled.

**Pull request closed without merging**
